gamemaster/game.php

```diff
@@ -1050,6 +1051,33 @@ protected function setWon(Member $Winner)
 		 */
 		global $DB;
 		$DB->sql_put("UPDATE wD_Games SET finishTime=".time()." WHERE id=".$this->id);
+		
+		//Do GR Stuff
+	 $SCcounts = array();
+	 $memberStatus = array();
+	 foreach($this->Members->ByStatus['Playing'] as $Member)
+	 {
+		 $SCcounts[$Member->userID] = $Member->supplyCenterNo;
+		 $memberStatus[$Member->userID] = 'Survived';
+	 }
+	 foreach($this->Members->ByStatus['Left'] as $Member)
+	 {
+		 $SCcounts[$Member->userID] = $Member->supplyCenterNo;
+		 $memberStatus[$Member->userID] = 'Resigned';
+	 }
+	 foreach($this->Members->ByStatus['Defeated'] as $Member)
+	 {
+		 $SCcounts[$Member->userID] = 0;
+		 $memberStatus[$Member->userID] = 'Defeated';
+	 }
+	 $memberStatus[$Winner->userID] = 'Won';

```

Comment:
This whole section of code is getting the status of each player. It pulls their status mainly from the $this->Members->ByStatus, but it doesn't exactly match up with what we want. Since $this->Members->ByStatus gives us arrays of varying sizes, we loop through those. However, the $Winner variable here is not an array, it is a single user, so we don't need to loop. That line overwrites the previous logic that sets all remaining players as survived, singling out the winner from the other surviving players.

Reply:
Is there a reason to not just pass the Members object to the GhostRatings instance? After the function calls Members->setWon() or Members->setDrawn() below, the member status and member SC count seem do be identical to ones you defined in your special GR arrays. So your "GR stuff" loops seem to be a bit redundant to me.

Reply:
I wrote this about 8 or 9 months ago, so I don't remember all the details, but I wrote this like this to make backfilling GR much easier. When backfilling GR, creating game objects or member objects was difficult. I tried a few ways and even talked to JMO about it, but the way that those objects are set up, they are difficult to create and get you to do what you want outside of the established game engine (this is also why we added the variant info table in the database). Thus, by having an array of statuses and supply center counts, it adds a few loops inside of the setWon and setDrawn functions, but creates a way for me to go straight from the database when backfilling GR. Also, if we removed that part of the code, we would have to extract from the members array inside of the GR calculations file anyway. 

So while it is probably possible to cut out this section and save 17 lines of code by passing in the members array, it would add more convoluted code in other areas.

Reply:
I understand.

My main concern are not the additional lines of code but having code at different places that does the same and is expected to stay the same. It could make further adjustments in the future more difficult as you need to remember to fix the (identical) code at any occurrence. Therefore I recommend using the existing status data of the Member objects instead of producing some new, i.e. directly use the status after the call Members->setDrawn()/setWon() (no matter how you pass it to the GR instance).

With the arrays needing to be provided in that way to the GR instance, let me suggest to use a different Members map instead:
After Members->setWon(), Members->setDrawn(), the following should in principle be equivalent to your code but uses the data calculated in the Members object and only one for loop:
`$SCcounts = array();`
`$memberStatus = array();`
`foreach($this->Members->ByOrder as $Member)`
`{`
`  $SCcounts[$Member->userID] = $Member->supplyCenterNo;`
`  $memberStatus[$Member->userID] = $Member->status;`
`}`
(not tested!)

This  7 lines would still be exactly identical setDrawn and setWon. For better maintainability you could consider to add an additional intermediate method that is called by both setWon and setDrawn to execute the identical GR code instead of having it in two places.

Reply:
Okay, having a prepGR function that does what you described would be useful. I will switch things around to include this in my next commit.

Reply:
Just put in a commit which addresses this.